From 72c52c3f0d195cde5571479f301469735afed42e Mon Sep 17 00:00:00 2001
From: Prasun Maiti <pmaiti@maxlinear.com>
Date: Wed, 30 Jun 2021 08:31:58 +0530
Subject: [PATCH] UGW_SW-60180 : net: dev: Move ppa s/w fastpath hook to
 process_backlog()

Currently, same skb is passing to s/w fastpath cb for logical interfaces on top
of base interface, which increases cpu overhead as there needs to parse header
again and match the session on hash lookup table.

Moved ppa s/w fastpath hook to process_backlog() to ensure skb is passed to cb
only for base interface.

Signed-off-by: Prasun Maiti <pmaiti@maxlinear.com>
---
 net/core/dev.c | 13 +++++--------
 1 file changed, 5 insertions(+), 8 deletions(-)

diff --git a/net/core/dev.c b/net/core/dev.c
index 8a5dd51292da..2dd89de1a4c1 100644
--- a/net/core/dev.c
+++ b/net/core/dev.c
@@ -4334,14 +4334,6 @@ static int __netif_receive_skb(struct sk_buff *skb)
 {
 	int ret;
 
-#if IS_ENABLED(CONFIG_PPA_API_SW_FASTPATH)
-	/* Try for PPA software acceleration, if possible */
-	if (ppa_hook_sw_fastpath_send_fn) {
-		if (ppa_hook_sw_fastpath_send_fn(skb) == 0)
-			return NET_RX_SUCCESS;
-	}
-#endif
-
 	if (sk_memalloc_socks() && skb_pfmemalloc(skb)) {
 		unsigned long pflags = current->flags;
 
@@ -4983,6 +4975,11 @@ static int process_backlog(struct napi_struct *napi, int quota)
 
 		while ((skb = __skb_dequeue(&sd->process_queue))) {
 			rcu_read_lock();
+#if IS_ENABLED(CONFIG_PPA_API_SW_FASTPATH)
+			/* Try for PPA software acceleration, if possible */
+			if (!ppa_hook_sw_fastpath_send_fn ||
+			    ppa_hook_sw_fastpath_send_fn(skb))
+#endif
 			__netif_receive_skb(skb);
 			rcu_read_unlock();
 			input_queue_head_incr(sd);
